src/models.jl

```diff
@@ -93,13 +84,14 @@ function Model(;
     tracers.S.data .= eos.S₀
     tracers.T.data .= eos.T₀
 
-    Model(metadata, configuration, boundary_conditions, constants, eos, grid,
-          velocities, tracers, pressures, G, Gp, forcing,
-          stepper_tmp, poisson_solver, clock, output_writers, diagnostics)
+    Model{typeof(arch)}(configuration, boundary_conditions, constants, eos, grid,
+                        velocities, tracers, pressures, G, Gp, forcing,
+                        stepper_tmp, poisson_solver, clock, output_writers, diagnostics)
 end
 
 "Legacy constructor for `Model`."
 Model(N, L; arch=:CPU, float_type=Float64) = Model(N=N, L=L; arch=arch, float_type=float_type)
 
+model_arch(model::Model{A}) where A <: Architecture = A
```

Comment:
Could just call this `arch`, since it only dispatches on `model` already.

Might be nice to have `float_type(m::Model) = eltype(model.grid)` and `eltype(::Grid{T}) where T = T` (or `RegularCartesianGrid` if `T` is not a type parameter of the abstract type `Grid`.

Reply:
Good point, just added both in!